keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
+//
+// Support: Core
+type ReferenceGrant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferenceGrant.
+	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+
+	// Note that `Status` sub-resource has been excluded at the
+	// moment as it was difficult to work out the design.
+	// A `Status` sub-resource may be added in the future.
+}
+
+// +kubebuilder:object:root=true
+// ReferenceGrantList contains a list of ReferenceGrant.
+type ReferenceGrantList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ReferenceGrant `json:"items"`
+}
+
+// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferenceGrantSpec struct {
+	// From describes the trusted namespaces and kinds that can reference the
+	// resources described in "To". Each entry in this list MUST be considered
+	// to be an additional place that references can be valid from, or to put
+	// this another way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	From []ReferenceGrantFrom `json:"from"`
+
+	// To describes the resources that may be referenced by the resources
+	// described in "From". Each entry in this list MUST be considered to be an
+	// additional place that references can be valid to, or to put this another
+	// way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	To []ReferenceGrantTo `json:"to"`
+}
+
+// ReferenceGrantFrom describes trusted namespaces and kinds.
+type ReferenceGrantFrom struct {
+	// Group is the group of the referent.
+	// When empty, the Kubernetes core API group is inferred.
+	Group Group `json:"group"`
+
+	// Kind is the kind of the referent.
+	Kind string `json:"kind"`
+
+	// Namespace is the namespace of the referent.
+	Namespace string `json:"namespace"`
+}
+
+// ReferenceGrantTo describes what Kinds are allowed as targets of the
+// references.
+type ReferenceGrantTo struct {
+	// Group is the group of the referent.
+	// When empty, the Kubernetes core API group is inferred.
+	Group string `json:"group"`
+
+	// Kind is the kind of the referent.
+	Kind string `json:"kind"`
+
+	// Name is the name of the referent. When unspecified, this policy
+	// refers to all resources of the specified Group and Kind in the local
+	// namespace.
+	//
+	// +optional
+	Name *string `json:"name,omitempty"`
```

Comment:
why name and not a selector?  Or a list of names?

Reply:
You can omit name to allow references to every resource of the specified Kind in a namespace. Cross-Namespace references are easy to mess up, I'd hate to further complicate this with label selectors. Not only would this complicate the implementation, I think it would result in confusing configuration. As far as a list of names, `To` is already a list, I'd rather avoid adding a list within a list.

Reply:
Yes, the design intent here is that separate rules are separate entries in the `To` list.

Reply:
In the case of VolumeSnapshots, a common usage pattern is to periodically take and rotate snapshots (every few hours, once a day, etc), where explicitly enumerating each would be cumbersome. With this model, to allow a single or subset of snapshots, some other automation would be required to rotate the names as the snapshots get rotated.

Reply:
Any should work for most use cases. My scenario of a single/subset was more hypothetical where someone may not want to expose all snapshots. And even so, they can express it with the proposal, it would just be some additional work to rotate the names.

Reply:
I suspect "any" would work, too, though we can certainly fabricate cases where a selector is needed.